scripts/anvil_tools/set_up_anvil_workspaces.py

```diff
@@ -0,0 +1,255 @@
+    """Add the anvil-admins@firecloud.org group as ADMIN to auth domain."""
+
+    admin_group_name = "anvil-admins@firecloud.org"
+
+    # request URL for addUserToGroup
+    uri = f"https://api.firecloud.org/api/groups/{auth_domain_name}/admin/{admin_group_name}"
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: */*" -H  "Authorization: Bearer [token]"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "*/*"}
+
+    # capture response from API and parse out status code
+    response = requests.put(uri, headers=headers)
+    status_code = response.status_code
+
+    # print success or fail message based on status code
+    if status_code == 204:
+        print(f"Successfully updated Authorization Domain, {auth_domain_name}, with group: {admin_group_name}.")
+
+    else:
+        print(f"WARNING: Failed to update Authorization Domain, {auth_domain_name}, with group: {admin_group_name}.")
+        print("Check output file for error details.")
+        print(response.json())
+
+
+def create_auth_domain(auth_domain_name):
+    """Create the Terra goolge group to be used as auth domain on workspaces."""
+
+    # request URL for createGroup
+    uri = f"https://api.firecloud.org/api/groups/{auth_domain_name}"
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: application/json" -H  "Authorization: Bearer [token]"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "application/json"}
+
+    # capture response from API and parse out status code
+    response = requests.post(uri, headers=headers)
+    status_code = response.status_code
+
+    # print success or fail message based on status code
+    if status_code == 201:
+        print(f"Successfully created Authorization Domain with name: {auth_domain_name}.")
+        return status_code
+    elif status_code in ["403", "409"]:
+        print(f"WARNING: Failed to create Authorization Domain with name: {auth_domain_name}.")
+        return "Authorization Domain with name already exists. Try again with unique group name."
+    else:
+        print(f"WARNING: Failed to create Authorization Domain with name: {auth_domain_name}.")
+        print("Check output file for error details.")
+        return(response.json())
+
+
+def setup_workspaces(tsv, project="anvil-datastorage"):
+    """Get the workspace and associated auth domain from input tsv file."""
+
+    # read full tsv into dataframe
+    setup_info_df = pd.read_csv(tsv, sep="\t")
+
+    # create df for output tsv file
+    col_names = ["input_workspace_name", "input_auth_domain_name",
+                 "auth_domain_email", "auth_domain_creation_error",
+                 "workspace_link", "workspace_creation_error",
+                 "workspace_ACLs", "workspace_ACLs_error",
+                 "workspace_setup_status"]
+    output_df = pd.DataFrame(columns=col_names)
+
+    # per row in tsv/df
+    for row in setup_info_df.index:
```

Comment:
it would probably be nice to refactor most of what's in this for loop into a `setup_workspace` function that does all the stuff for a single workspace - that's the core of the work that's happening, and `setup_workspaces` is really just a wrapper to apply this to all the workspaces/rows in the input file

Reply:
the function would take in all workspace-related inputs, and it would return all the info needed to write to your status output file